logging.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+import Control.Applicative ((<|>))
+import Control.Exception.Safe (catchAny, displayException)
+import Control.Monad (guard)
+import Data.Foldable (fold)
+import System.Directory (doesFileExist)
+import System.IO (hPutStrLn, stderr)
+
+
+data Log =
+  Log {
+    logInfo :: String -> IO (),
+    logError :: String -> IO ()
+  }
+
+
+-- A log that writes to the console.
+consoleLog =
+  Log {
+    logInfo = putStrLn,
+    logError = hPutStrLn stderr
+  }
+
+
+-- A log that writes to the specified files.
+fileLog infoPath errPath =
+  Log {
+    logInfo = appendFile infoPath . (<> "\n"),
+    logError = appendFile errPath . (<> "\n")
```

Comment:
What's the default BufferMode for appendFile? Should I force this to flush, or otherwise change the buffer mode? Or is it fine?

Reply:
It uses `withFile`, which brackets the writing with opening and closing a file handle, so it seems like it should be fine?

Reply:
I assume that `appendFile` closes its file handle immediately after writing, which would flush, so surely this is fine.

Reply:
I'm mostly asking because I'm slightly afraid of someone coming back five years from now saying, "see, friedbrice doesn't know how to properly use `appendFile`" 🙃